RenderStateNotationParser/src/RenderStateNotationParserImpl.cpp

```diff
@@ -178,133 +202,197 @@ RenderStateNotationParserImpl::RenderStateNotationParserImpl(IReferenceCounters*
-
-            if (Includes.find(Path) == Includes.end())
-            {
-                Includes.insert(Path);
-                ParseJSON({Path.c_str(), nullptr, ParserCI.pStreamFactory});
-            }
-        }
-
-        for (auto const& Shader : Json["Shaders"])
+        else if (Json.is_object())
         {
             ShaderCreateInfo ResourceDesc = {};
-            Deserialize(Shader, ResourceDesc, *m_pAllocator);
+            Deserialize(Json, ResourceDesc, Allocator);
             VERIFY_EXPR(ResourceDesc.Desc.Name != nullptr);
+
             m_ShaderNames.emplace(HashMapStringKey{ResourceDesc.Desc.Name, false}, StaticCast<Uint32>(m_Shaders.size()));
             m_Shaders.push_back(ResourceDesc);
+            Name = ResourceDesc.Desc.Name;
         }
+        else
+        {
+            throw nlohmann::json::type_error::create(302, std::string("type must be object or string, but is ") + Json.type_name(), Json);
+        }
+    };
 
-        for (auto const& RenderPass : Json["RenderPasses"])
+    Callbacks.RenderPassCallback = [this](const nlohmann::json& Json, const char*& Name, DynamicLinearAllocator& Allocator) {
+        if (Json.is_string())
+        {
+            Deserialize(Json, Name, Allocator);
+        }
+        else if (Json.is_object())
         {
             RenderPassDesc ResourceDesc = {};
-            Deserialize(RenderPass, ResourceDesc, *m_pAllocator);
+            Deserialize(Json, ResourceDesc, Allocator);
             VERIFY_EXPR(ResourceDesc.Name != nullptr);
+
             m_RenderPassNames.emplace(HashMapStringKey{ResourceDesc.Name, false}, StaticCast<Uint32>(m_RenderPasses.size()));
             m_RenderPasses.push_back(ResourceDesc);
+            Name = ResourceDesc.Name;
+        }
+        else
+        {
+            throw nlohmann::json::type_error::create(302, std::string("type must be object or string, but is ") + Json.type_name(), Json);
         }
+    };
 
-        for (auto const& Signature : Json["ResourceSignatures"])
+    Callbacks.ResourceSignatureCallback = [this](const nlohmann::json& Json, const char*& Name, DynamicLinearAllocator& Allocator) {
+        if (Json.is_string())
+        {
+            Deserialize(Json, Name, Allocator);
+        }
+        else if (Json.is_object())
         {
             PipelineResourceSignatureDesc ResourceDesc = {};
-            Deserialize(Signature, ResourceDesc, *m_pAllocator);
+            Deserialize(Json, ResourceDesc, Allocator);
             VERIFY_EXPR(ResourceDesc.Name != nullptr);
+
             m_ResourceSignatureNames.emplace(HashMapStringKey{ResourceDesc.Name, false}, StaticCast<Uint32>(m_ResourceSignatures.size()));
             m_ResourceSignatures.push_back(ResourceDesc);
+            Name = ResourceDesc.Name;
+        }
+        else
+        {
+            throw nlohmann::json::type_error::create(302, std::string("type must be object or string, but is ") + Json.type_name(), Json);
         }
+    };
```

Comment:
There is quite a bit of copy-paste here. The logic is repeated 3 times with only difference that it is applied to Shaders, Render Passes or Signatures. Try using auto in lambas to write one function that will handle all three cases.

Reply:
We'll need this in the future for implicit and default states